integrations: Add links and topics to Taiga webhook message.

orientor opened the pull request:
In the current Taiga Integration webhook messages look like:

> User did this
But Taiga webhook got upgraded and now provides link support as well. So use the permalink in the template if available so that it now looks like:

> [User](www.google.com) did [this](www.google.com).

Otherwise show the old Taiga messge as support for the older Taiga version has also been preserved.

To do this, change all the templates and create new functions which check the webhook JSON and 
return the username, issue name, subject name, etc. based on the link availabilty in JSON file.

Also add more events which were not covered in the previous Taiga implimentation.

If some activity happens under a particular 'sprint' in Taiga then publish the webhook message of that activity under the 'sprint name' topic, which is available in the webhook JSON, so that it is more clear to the users.

Change the tests accordingly and added new tests for the events that the current Taiga integration was lacking and also for old events with new added links.

Fixes #13698 .

timabbott commented:
@orientor it looks like you've accidentally introduced some merge commits, which is going to create all sorts of pain for you.  

Check out our [GitHub guide](https://zulip.readthedocs.io/en/latest/git-guide.html) and [commit guidelines](https://zulip.readthedocs.io/en/latest/contributing/version-control.html) for more details.  you probably want either to be using `git pull --rebase` or to set the Git config option that makes `git pull` do rebases by default.

Also, the tests are failing; you should be able to debug efficiently with `test-backend zerver/webhooks/taiga`.

timabbott commented on zerver/webhooks/taiga/view.py:
```diff
@@ -23,217 +22,237 @@
     for event in parsed_events:
         content_lines.append(generate_content(event) + '\n')
-    content = "".join(sorted(content_lines))
+    content = ''.join(sorted(content_lines))
```
Why did you change all the `""` to `''` in this file?  It makes your changes very hard to read, and is not in line with the conventions of the codebase.  Can you please undo all the irrelevant changes where all you did was changed the quote format?

timabbott commented on zerver/webhooks/taiga/view.py:
```diff
@@ -242,61 +261,64 @@ def parse_change_event(change_type: str, message: Mapping[str, Any]) -> Optional
+    data = message['data']
+    if 'permalink' in data:
+        return '['+data.get('subject', data.get('name'))+']'+'('+data['permalink']+')'
+    return '**'+data.get('subject', data.get('name'))+'**'
```
Please include a space on either side of each `+`; that's standard Python style that helps with readability.

timabbott commented on zerver/webhooks/taiga/view.py:
```diff
@@ -242,61 +261,64 @@ def parse_change_event(change_type: str, message: Mapping[str, Any]) -> Optional
+
+def get_epic_subject(message: Mapping[str, Any]) -> str:
+    if 'permalink' in message['data']['epic']:
+        return'['+message['data']['epic']['subject']+']'+'('+message['data']['epic']['permalink']+')'
```
Also, please include a space after `return`; it's certainly bad style (I'm surprised it even works without the space).

timabbott commented on zerver/webhooks/taiga/view.py:
```diff
@@ -242,61 +261,64 @@ def parse_change_event(change_type: str, message: Mapping[str, Any]) -> Optional
             events.append(parse_comment(message))
-    elif message["action"] == "test":
+    elif message['action'] == 'test':
         events.append(parse_webhook_test(message))
```
It looks like you wouldn't have needed to change this function at all were you not reworking the quote formatting.

orientor commented:
@timabbott updated the commit accordingly.

Changed the commit message and updated the PR description so that its now clear as to what I am doing.

The quotation style changes have been rolled back. I was actually using it for easing development and meant to roll it back on pushing but forgot about it. Due to this some functions which I did not change at all were also being show as edited.

Also made code more readable by adding spaces. Did not expect this as the linter had previously caught such errors in the same file. Also linter did not catch no space after return. But all has been fixed now.

orientor commented:
@timabbott I think it should be mergeable now. Kindly check.

timabbott commented:
@orientor I merged this as 7d2bb707e785ea5c75e1340d1e8ea60250b456a8, after changing the commit message to be clear about what it does and not close #13698.

It shouldn't close that issue because you haven't fixed it -- we have links to the users, which is useful, but the critical links to include here are the permanent links to the issues/epics/etc., and this PR doesn't do that.  Can you submit a follow-up PR adding those links as well?  Thanks @orientor!

orientor commented on zerver/webhooks/taiga/view.py:
```diff
@@ -278,25 +298,28 @@ def generate_content(data: Mapping[str, Any]) -> str:
     data = message["data"]
-    return data.get("subject", data.get("name"))
+    if 'permalink' in data:
+        return '[' + data.get('subject', data.get('name')) + ']' + '(' + data['permalink'] + ')'
```
@timabbott Sorry I didn't mention it in the PR clearly but the PR actually adds links to issues/ epics etc. I have added a few tests for them and also the role of the functions at the end of views.py file is to use issue/sprint name with link if possible else just use the name. The code which I have highlighted is one of the functions.
